src/parser.js

```diff
@@ -667,11 +675,54 @@ function mapBlockStatement(node, meta, factory = b.blockStatement) {
 }
 
 function mapConditionalAssignment({variable, value, context, operator}, meta) {
-  let memberExpression = b.identifier(variable.base.value);
+  const {condition, memberExpression} = mapSoakedChainToConditions(variable);
+
+  return b.ifStatement(
+    condition,
+    b.blockStatement(
+      [
+        b.expressionStatement(
+          b.assignmentExpression(context || operator || '=', memberExpression, mapExpression(value, meta))
+        ),
+      ]
+    )
+  );
+}
+
+function mapConditionalUpdateOperator({first: value, context, operator, flip}) {
+  const {condition, memberExpression} = mapSoakedChainToConditions(value);
+
+  return b.ifStatement(
+    condition,
+    b.blockStatement(
+      [
+        b.expressionStatement(
+          b.updateExpression(context || operator, memberExpression, !flip)
+        ),
+      ]
+    )
+  );
+}
+
+function nodeHasProperties(node) {
+  return node && node.properties && node.properties.length > 0;
+}
+
+function mapSoakedChainToConditions({properties, base}) {
+  let memberExpression = b.identifier(base.value);
   let condition;
 
-  for (const property of variable.properties) {
-    const identifier = b.identifier(property.name.value);
+  for (const property of properties) {
+    let memberProperty;
+    const isArrayAccessor = (property.constructor.name === 'Index');
+
+    if (nodeHasProperties(property.index)) {
```

Comment:
@lemonmade managed to clean this up with `mapLiteral`.  Not perfect, but combined with your suggestions, it's way more readable.